File: src/tag/diff_context.py
"""PRD-038: Diff-Aware Context Injection (tag context inject --git-diff).

Scopes context injection to only the files changed in a git diff.
Injects as a user-message turn (never modifies system prompt) so the
system prompt stays eligible for prompt caching at 0.1x cost.
"""
from __future__ import annotations

import fnmatch
import subprocess
from pathlib import Path
from typing import Iterator

# Files always blocked from injection
DEFAULT_BLOCKED_PATTERNS = [
    ".env", "*.env", ".env.*",
    "*.key", "*.pem", "*.p12", "*.pfx",
    "*secret*", "*credential*", "*password*",
    "*.token",
]

_BINARY_EXTS = {
    ".png", ".jpg", ".jpeg", ".gif", ".ico", ".svg", ".webp",
    ".zip", ".tar", ".gz", ".bz2", ".xz", ".7z", ".rar",
    ".exe", ".dll", ".so", ".dylib", ".pyc",
    ".pdf", ".docx", ".xlsx", ".pptx",
    ".mp4", ".mov", ".mp3", ".wav",
    ".ttf", ".woff", ".woff2", ".eot",
}

WARN_TOKEN_THRESHOLD = 10_000


def _is_blocked(filename: str, blocked_patterns: list[str]) -> bool:
    name = Path(filename).name
    for pat in blocked_patterns:
        if fnmatch.fnmatch(name, pat) or fnmatch.fnmatch(filename, pat):
            return True
    return False


def _is_binary(filename: str) -> bool:
    return Path(filename).suffix.lower() in _BINARY_EXTS


def get_changed_files(
    ref: str = "HEAD",
    *,
    staged: bool = False,
    workdir: Path | None = None,
) -> list[str]:
    """Return list of changed file paths from git diff."""
    cmd = ["git", "diff", "--name-only"]
    if staged:
        cmd.append("--cached")
    else:
        cmd.append(ref)
    try:
        result = subprocess.run(
            cmd, capture_output=True, text=True,
            cwd=str(workdir or Path.cwd()),
        )
        if result.returncode != 0:
            raise RuntimeError(f"git diff failed: {result.stderr.strip()}")
        files = [f.strip() for f in result.stdout.splitlines() if f.strip()]
        return files
    except FileNotFoundError:
        raise RuntimeError("git not found in PATH")


def get_file_diff(
    filename: str,
    ref: str = "HEAD",
    *,
    context_lines: int = 3,
    staged: bool = False,
    workdir: Path | None = None,
) -> str:
    """Return unified diff for a single file."""
    cmd = ["git", "diff", f"-U{context_lines}"]
    if staged:
        cmd.append("--cached")
    else:
        cmd.append(ref)
    cmd.extend(["--", filename])

    try:
        result = subprocess.run(
            cmd, capture_output=True, text=True,
            cwd=str(workdir or Path.cwd()),
        )
        return result.stdout
    except Exception:
        return ""


def _estimate_tokens(text: str) -> int:
    """Rough token estimate: ~4 chars per token."""
    return max(1, len(text) // 4)


def build_diff_context(
    ref: str = "HEAD",
    *,
    staged: bool = False,
    context_lines: int = 3,
    max_files: int = 10,
    blocked_patterns: list[str] | None = None,
    workdir: Path | None = None,
) -> dict:
    """Build a diff context block for injection.

    Returns:
        {
            "content": str,          # the assembled diff text
            "files_included": list,
            "files_skipped": list,
            "estimated_tokens": int,
            "warn": bool,            # True if > WARN_TOKEN_THRESHOLD
        }
    """
    patterns = (blocked_patterns or []) + DEFAULT_BLOCKED_PATTERNS

    all_changed = get_changed_files(ref, staged=staged, workdir=workdir)
    included: list[str] = []
    skipped: list[str] = []

    for f in all_changed:
        if _is_blocked(f, patterns):
            skipped.append(f)
            continue
        if _is_binary(f):
            skipped.append(f)
            continue
        if len(included) >= max_files:
            skipped.append(f)
            continue
        included.append(f)

    diff_parts: list[str] = []
    for f in included:
        diff_text = get_file_diff(f, ref, context_lines=context_lines, staged=staged, workdir=workdir)
        if diff_text.strip():
            diff_parts.append(f"### {f}\n```diff\n{diff_text.rstrip()}\n```")

    content = "\n\n".join(diff_parts)
    tokens = _estimate_tokens(content)

    return {
        "content": content,
        "files_included": included,
        "files_skipped": skipped,
        "estimated_tokens": tokens,
        "warn": tokens > WARN_TOKEN_THRESHOLD,
    }


def pr_diff_context(
    pr_number: int | str,
    repo: str | None = None,
    *,
    context_lines: int = 3,
    max_files: int = 10,
    blocked_patterns: list[str] | None = None,
) -> dict:
    """Fetch and filter a GitHub PR diff via gh CLI."""
    patterns = (blocked_patterns or []) + DEFAULT_BLOCKED_PATTERNS
    cmd = ["gh", "pr", "diff", str(pr_number), "--patch"]
    if repo:
        cmd.extend(["--repo", repo])
    try:
        result = subprocess.run(cmd, capture_output=True, text=True)
        if result.returncode != 0:
            raise RuntimeError(f"gh pr diff failed: {result.stderr.strip()}")
        raw_diff = result.stdout
    except FileNotFoundError:
        raise RuntimeError("gh CLI not found. Install GitHub CLI.")

    # Parse filenames from diff headers
    included: list[str] = []
    skipped: list[str] = []
    current_file: str | None = None
    file_diffs: dict[str, list[str]] = {}

    for line in raw_diff.splitlines():
        if line.startswith("diff --git "):
            parts = line.split(" b/")
            if len(parts) >= 2:
                current_file = parts[-1].strip()
                if _is_blocked(current_file, patterns) or _is_binary(current_file):
                    skipped.append(current_file)
                    current_file = None
                elif len(included) >= max_files:
                    skipped.append(current_file)
                    current_file = None
                else:
                    included.append(current_file)
                    file_diffs[current_file] = [line]
        elif current_file and current_file in file_diffs:
            file_diffs[current_file].append(line)

    diff_parts = []
    for f in included:
        text = "\n".join(file_diffs.get(f, []))
        if text.strip():
            diff_parts.append(f"### {f}\n```diff\n{text}\n```")

    content = "\n\n".join(diff_parts)
    tokens = _estimate_tokens(content)
    return {
        "content": content,
        "files_included": included,
        "files_skipped": skipped,
        "estimated_tokens": tokens,
        "warn": tokens > WARN_TOKEN_THRESHOLD,
    }

File: tag-go/internal/diffcontext/diffcontext.go
// Package diffcontext builds a git-diff context block for agent injection
// (port of src/tag/diff_context.py). Secret/binary files are filtered out and
// the total is token-estimated. Shells out to the local `git` binary.
package diffcontext

import (
	"os/exec"
	"path/filepath"
	"strings"
)

// DefaultBlockedPatterns are always excluded (secrets/credentials).
var DefaultBlockedPatterns = []string{
	".env", "*.env", ".env.*",
	"*.key", "*.pem", "*.p12", "*.pfx",
	"*secret*", "*credential*", "*password*",
	"*.token",
}

var binaryExts = map[string]bool{
	".png": true, ".jpg": true, ".jpeg": true, ".gif": true, ".ico": true, ".svg": true, ".webp": true,
	".zip": true, ".tar": true, ".gz": true, ".bz2": true, ".xz": true, ".7z": true, ".rar": true,
	".exe": true, ".dll": true, ".so": true, ".dylib": true, ".pyc": true,
	".pdf": true, ".docx": true, ".xlsx": true, ".pptx": true,
	".mp4": true, ".mov": true, ".mp3": true, ".wav": true,
	".ttf": true, ".woff": true, ".woff2": true, ".eot": true,
}

// WarnTokenThreshold triggers the "large diff" warning.
const WarnTokenThreshold = 10000

// Result is the assembled diff context.
type Result struct {
	Content         string   `json:"content"`
	FilesIncluded   []string `json:"files_included"`
	FilesSkipped    []string `json:"files_skipped"`
	EstimatedTokens int      `json:"estimated_tokens"`
	Warn            bool     `json:"warn"`
}

func isBlocked(filename string, patterns []string) bool {
	name := filepath.Base(filename)
	for _, pat := range patterns {
		if ok, _ := filepath.Match(pat, name); ok {
			return true
		}
		if ok, _ := filepath.Match(pat, filename); ok {
			return true
		}
	}
	return false
}

func isBinary(filename string) bool {
	return binaryExts[strings.ToLower(filepath.Ext(filename))]
}

func estimateTokens(text string) int {
	t := len(text) / 4
	if t < 1 {
		return 1
	}
	return t
}

// GitError is returned for git failures with a concise message.
type GitError struct{ Msg string }

func (e *GitError) Error() string { return e.Msg }

func validateRef(ref string) error {
	if strings.HasPrefix(ref, "-") {
		return &GitError{Msg: "invalid git ref: " + ref}
	}
	return nil
}

func changedFiles(ref string, staged bool, workdir string) ([]string, error) {
	args := []string{"diff", "--name-only"}
	if staged {
		args = append(args, "--cached")
	} else {
		if err := validateRef(ref); err != nil {
			return nil, err
		}
		args = append(args, ref)
	}
	cmd := exec.Command("git", args...)
	cmd.Dir = workdir
	out, err := cmd.CombinedOutput()
	if err != nil {
		if _, ok := err.(*exec.Error); ok {
			return nil, &GitError{Msg: "git not found in PATH"}
		}
		low := strings.ToLower(string(out))
		if strings.Contains(low, "not a git repository") || strings.Contains(low, "usage:") || strings.TrimSpace(string(out)) == "" {
			return nil, &GitError{Msg: "not a git repository (or no commits to diff against)"}
		}
		first := strings.SplitN(strings.TrimSpace(string(out)), "\n", 2)[0]
		return nil, &GitError{Msg: "git diff failed: " + first}
	}
	var files []string
	for _, f := range strings.Split(string(out), "\n") {
		if s := strings.TrimSpace(f); s != "" {
			files = append(files, s)
		}
	}
	return files, nil
}

func fileDiff(filename, ref string, contextLines int, staged bool, workdir string) string {
	args := []string{"diff", "-U" + itoa(contextLines)}
	if staged {
		args = append(args, "--cached")
	} else {
		if validateRef(ref) != nil {
			return ""
		}
		args = append(args, ref)
	}
	args = append(args, "--", filename)
	cmd := exec.Command("git", args...)
	cmd.Dir = workdir
	out, err := cmd.Output()
	if err != nil {
		return ""
	}
	return string(out)
}

func itoa(n int) string {
	if n == 0 {
		return "0"
	}
	neg := n < 0
	if neg {
		n = -n
	}
	var b []byte
	for n > 0 {
		b = append([]byte{byte('0' + n%10)}, b...)
		n /= 10
	}
	if neg {
		b = append([]byte{'-'}, b...)
	}
	return string(b)
}

// Build assembles the diff context per the ref/staged mode, filtering blocked
// and binary files and capping at maxFiles.
func Build(ref string, staged bool, contextLines, maxFiles int, extraBlocked []string, workdir string) (*Result, error) {
	patterns := append(append([]string{}, extraBlocked...), DefaultBlockedPatterns...)
	all, err := changedFiles(ref, staged, workdir)
	if err != nil {
		return nil, err
	}
	var included, skipped []string
	for _, f := range all {
		if isBlocked(f, patterns) || isBinary(f) || len(included) >= maxFiles {
			skipped = append(skipped, f)
			continue
		}
		included = append(included, f)
	}
	var parts []string
	for _, f := range included {
		d := fileDiff(f, ref, contextLines, staged, workdir)
		if strings.TrimSpace(d) != "" {
			parts = append(parts, "### "+f+"\n```diff\n"+strings.TrimRight(d, "\n")+"\n```")
		}
	}
	content := strings.Join(parts, "\n\n")
	tokens := estimateTokens(content)
	return &Result{
		Content: content, FilesIncluded: included, FilesSkipped: skipped,
		EstimatedTokens: tokens, Warn: tokens > WarnTokenThreshold,
	}, nil
}
